### PR TITLE
[quant][graph] Add support for FP16 dynamic quant

### PR DESCRIPTION
**Pull request by supriyar**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42348 [quant] Use PlaceholderObserver instead of Fp16Observer and NoopObserver
* **#42222 [quant][graph] Add support for FP16 dynamic quant**
* #42221 [quant] Add FP16Observer for fp16 quant support
* #42147 [quant] Add saturate_to_fp16 op for FP16 quant support

Summary:
This change adds the necessary passes to perform FP16 dynamic quantization.
We skip inserting observers for activations based on the dtype (torch.float16) and only insert the Fp16Observer for weights

Test Plan:
python test/test_quantization.py TestQuantizeJitOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22849220](https://our.internmc.facebook.com/intern/diff/D22849220)

### CODE REVIEW

File: torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp [hunk 1]
```diff
@@ -405,7 +394,7 @@ void insertQuantizationOps(
     return;
   }
   if (quant_type == QuantType::DYNAMIC) {
-    if (isFP16NoopObserver(module, observer)) {
+    if (isFp16Observer(observer->input(0))) {
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:397]: can we just check dtype here?

File: torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp [hunk 2]
```diff
@@ -245,12 +232,14 @@ bool isNoopObserver(Value* observer) {
   return false;
 }
 
-bool isFP16NoopObserver(script::Module& module, Node* n) {
-  Value* v = n->output();
-  auto observer = n->input(0);
-  auto observer_module = module.attr(findObserverName(v).value()).toModule();
-  return (observer_module.attr("dtype") == at::ScalarType::Half) &&
-      isNoopObserver(observer);
+bool isFp16Observer(Value* observer) {
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:235]: do we need this check? I think checking dtype is enough for our purposes?

File: torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp (hunk 1 above)
Comment (supriyar) [reply torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:397]: I think so, was just being more explicit by checking for the observer type as well.

File: torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp [hunk 3]
```diff
@@ -405,7 +394,9 @@ void insertQuantizationOps(
     return;
   }
   if (quant_type == QuantType::DYNAMIC) {
-    if (isFP16NoopObserver(module, observer)) {
+    if (getObserverDtype(module, observer_out).has_value() &&
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:397]: nit: probably better to assign the result to a variable

File: torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp [hunk 4]
```diff
@@ -245,12 +232,14 @@ bool isNoopObserver(Value* observer) {
   return false;
 }
 
-bool isFP16NoopObserver(script::Module& module, Node* n) {
-  Value* v = n->output();
-  auto observer = n->input(0);
-  auto observer_module = module.attr(findObserverName(v).value()).toModule();
-  return (observer_module.attr("dtype") == at::ScalarType::Half) &&
-      isNoopObserver(observer);
+c10::optional<at::ScalarType> getObserverDtype(Module& module, Value* v) {
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization/insert_quant_dequant.cpp:235]: also maybe we can return at::ScalarType?

File: test/quantization/test_quantize_jit.py [hunk 5]
```diff
@@ -2506,20 +2506,22 @@ def forward(self, x):
...
-            assert len(attrs_with_prefix(m, '_observer_')) == 1
-            # for weight
+            # observer for weight
             assert len(attrs_with_prefix(m.fc, '_observer_')) == 1
+
             if qconfig == float16_dynamic_qconfig:
-                observer_name = 'NoopObserver = prim::GetAttr[name="_observer_'
+                observer_name = 'Fp16Observer = prim::GetAttr[name="_observer_'
+                FileCheck().check(observer_name) \
+                           .run(m.fc.graph)
```
Comment (jerryzh168) [inline test/quantization/test_quantize_jit.py:2515]: Looks like this check is not very useful, what do we want to check here?

**Review by jerryzh168 — APPROVED** [review]
LG, had a few inline comments

File: test/quantization/test_quantize_jit.py (hunk 5 above)
Comment (supriyar) [reply test/quantization/test_quantize_jit.py:2515]: Just additional check for observer name match

**Pull request closed without merging**